examples/IB/explicit/ex7/example.cpp

```diff
@@ -0,0 +1,435 @@
+        }
+        // Write out initial visualization data.
+        int iteration_num = time_integrator->getIntegratorStep();
+        double loop_time = time_integrator->getIntegratorTime();
+        if (dump_viz_data && uses_visit)
+        {
+            pout << "\n\nWriting visualization files...\n\n";
+            time_integrator->setupPlotData();
+            visit_data_writer->writePlotData(patch_hierarchy, iteration_num, loop_time);
+            silo_data_writer->writePlotData(iteration_num, loop_time);
+        }
+        
+        // Create Text File which Stores the locations of each IB point
+
+        ofstream fout("Coordinates.txt", ios::out);
+
+        // Main time step loop.
+        double loop_time_end = time_integrator->getEndTime();
+        double dt = 0.0;
+        while (!MathUtilities<double>::equalEps(loop_time, loop_time_end) && time_integrator->stepsRemaining())
+        {
+            iteration_num = time_integrator->getIntegratorStep();
+            loop_time = time_integrator->getIntegratorTime();
+
+            pout << "\n";
+            pout << "+++++++++++++++++++++++++++++++++++++++++++++++++++\n";
+            pout << "At beginning of timestep # " << iteration_num << "\n";
+            pout << "Simulation time is " << loop_time << "\n";
+
+
+            update_target_points(patch_hierarchy, ib_method_ops->getLDataManager(), loop_time, dt);
+
+
+            dt = time_integrator->getMaximumTimeStepSize();
+            time_integrator->advanceHierarchy(dt);
+            print_eul_data(patch_hierarchy, ib_method_ops->getLDataManager());
+            loop_time += dt;
+
+            pout << "\n";
+            pout << "At end       of timestep # " << iteration_num << "\n";
+            pout << "Simulation time is " << loop_time << "\n";
+            pout << "+++++++++++++++++++++++++++++++++++++++++++++++++++\n";
+            pout << "\n";
+
+            std::ofstream fout("Coordinates.txt", std::ios::app);
+
+            const int finest_ln = patch_hierarchy->getFinestLevelNumber();
+            Pointer<LData> X_data = ib_method_ops->getLDataManager()->getLData("X", finest_ln);
+            Vec X_vec = X_data->getVec();
+            double* x_vals;
+            int ierr = VecGetArray(X_vec, &x_vals);
+            IBTK_CHKERRQ(ierr);
+            Pointer<LMesh> l_mesh = ib_method_ops->getLDataManager()->getLMesh(finest_ln);
+            const std::vector<LNode*>& local_nodes = l_mesh->getLocalNodes();
+            for (const auto& node : local_nodes)
+            {
+                const int lag_idx = node->getLagrangianIndex();
+                const int petsc_idx = node->getLocalPETScIndex();
+                Eigen::Map<Vector3d> X(&x_vals[petsc_idx * NDIM]);
+                fout.unsetf(ios_base::showpos);
+                fout.setf(ios_base::scientific);
+                fout.precision(5);
+                fout << loop_time << "," << X(0) << "," << X(1) << "," << X(2) << "\n";
+            }
+
```

Comment:
We need a matching call to `VecRestoreArray()` here.

Reply:
For consistency we should check PETSc's return code.

History lesson: PETSc functions always return an integer indicating whether or not they succeeded (success is `0`, anything else is some sort of failure). Traditionally this is named `ierr` because in the old days of FORTRAN-77, variables starting with `i`,`j`, ... `n` were implicitly integers and those starting with everything else were implicitly floats. This function isn't going to fail but for the sake of good style we should never ignore the return code - we can use `IBTK_CHKERRQ` again down here to match the first PETSc call.